File: 7_k_fold.py
from sklearn.linear_model import LinearRegression
from math import sqrt # We'll need sqrt()
import statistics # for mean()
from sklearn.metrics import mean_squared_error # for mean_squared_error which calculates 
from sklearn.model_selection import KFold # import KFold
from sklearn.preprocessing import PolynomialFeatures
import numpy as np
import pandas as pd


# Read the data
hhData = pd.read_csv("forestfires.csv", header=0, sep=",", engine='python')

#sampling
#hhData = hhData.sample(frac=1).reset_index(drop=True)

kf = KFold(n_splits=10)

# Create an empty array where we will store the calculated RMSE values
# so that we may be able to 
allRMSE = np.empty(shape=[0, 1])

# Just a variable to count at which tests we are
testNumber = 0

for train_index, test_index in kf.split(hhData):

 # Next test
 testNumber += 1
 
 # Use the current indexes train_index and test_index to get the actual observations for the
 # training and testing of the model respectively.

 # From the original, complete dataset get the data with which we will use TRAIN our model (aka the training set),
 # i.e. estimate the coefficients. We get the rows designated by train_index and all their columns/variables
 trainingData = hhData.iloc[train_index,:]

 # From the original, complete dataset get the data with which we will TEST our model, i.e. estimate its
 # accuracy. This is the "unknown" dataset.
 # Note: we do know the value of the dependent variable FoodExpenditure for the testing set
 # and hence we will be able to estimate the prediction error/accuracy
 testData = hhData.iloc[test_index,:]

 # Use the training data to estimate the coefficients of the multiple linear regression model.
 # The model we will estimate is: FoodExpenditure = b1Income + b2FamilySize + b0
 lm = LinearRegression(normalize=False, fit_intercept=True)

 # Since the linear regression model has as independent variables Income and FamilySize,
 # we get these variables from the training dataset.
 # The method .fit() does the actual estimation of the coefficients for the linear regression
 # model using OLS (Ordinary Least Squares).
 # The first argument of the .fit() method are the values of the independent variables -in our case
 # trainingData.loc[:,['Income','FamilySize']] - and the second
 # argument are the values of the dependent variable, here trainingData.loc[:,['FoodExpenditure']].
 estimatedModel = lm.fit(trainingData.loc[:,['temp','wind','rain']], trainingData.loc[:,['area']])

 # Coefficients have been estimated. Take a look at them. Not that it is important but heck, why not.
 # NOTE: the coefficients are returned (.coef_) as an array. The two numbers you'll see when
 # printing .coef_  must be interpreted as follows: first number is the coefficient
 # for Income, second number the coefficient for FamilySize. The estimated constant term b0 can retrieved 
 # via the .intercept_ variable.
 print(">>>Iteration ", testNumber, sep='')
 print("\tEstimated coefficients:")
 print("\t\tb1=", estimatedModel.coef_[0][0] , sep='')
 print("\t\tb2=", estimatedModel.coef_[0][1] , sep='') 
 print("\t\tb0=", estimatedModel.intercept_, sep='')
    
 predictedExpenditure = estimatedModel.predict(testData.loc[:,['temp','wind','rain']])

 # Calculate the Root Mean Squared Error (RMSE) for this testing set.
 # We have the real values of the dependent variable from the original dataset
 # which is testData.loc[:,['FoodExpenditure']] and the model's predicted values in predictedExpenditure
 # IMPORTANT: Please note the following: the function mean_squared_error() calculate the MSE NOT the RMSE. In
 # order to get the RMSE, we need to square the value returned by mean_squared_error(). See your notes on
 # how MSE and RMSE differ.
 RMSE = sqrt(mean_squared_error(testData.loc[:,['area']], predictedExpenditure))

 # Display the RMSE value
 print("\t\tModel RMSE=", RMSE, sep='')
 
 # Also, store the calculated RMSE value into an array, so that we can calculate the mean error after k-fold cross
 # validation has been finished 
 allRMSE = np.append(allRMSE, RMSE)

testNumber = 0



print("\n=======================================================")
print(" Final result: Mean RMSE of tests:", statistics.mean(allRMSE), sep='' )
print("=======================================================")

#filtering for area < 3.2
hhData_2 = hhData.loc[hhData['area'] < 3.2]

for train_index, test_index in kf.split(hhData_2):

 # Next test
 testNumber += 1
 
 # Use the current indexes train_index and test_index to get the actual observations for the
 # training and testing of the model respectively.

 # From the original, complete dataset get the data with which we will use TRAIN our model (aka the training set),
 # i.e. estimate the coefficients. We get the rows designated by train_index and all their columns/variables
 trainingData = hhData.iloc[train_index,:]

 # From the original, complete dataset get the data with which we will TEST our model, i.e. estimate its
 # accuracy. This is the "unknown" dataset.
 # Note: we do know the value of the dependent variable FoodExpenditure for the testing set
 # and hence we will be able to estimate the prediction error/accuracy
 testData = hhData.iloc[test_index,:]

 # Use the training data to estimate the coefficients of the multiple linear regression model.
 # The model we will estimate is: FoodExpenditure = b1Income + b2FamilySize + b0
 lm = LinearRegression(normalize=False, fit_intercept=True)

 # Since the linear regression model has as independent variables Income and FamilySize,
 # we get these variables from the training dataset.
 # The method .fit() does the actual estimation of the coefficients for the linear regression
 # model using OLS (Ordinary Least Squares).
 # The first argument of the .fit() method are the values of the independent variables -in our case
 # trainingData.loc[:,['Income','FamilySize']] - and the second
 # argument are the values of the dependent variable, here trainingData.loc[:,['FoodExpenditure']].
 estimatedModel = lm.fit(trainingData.loc[:,['temp','wind','rain']], trainingData.loc[:,['area']])

 # Coefficients have been estimated. Take a look at them. Not that it is important but heck, why not.
 # NOTE: the coefficients are returned (.coef_) as an array. The two numbers you'll see when
 # printing .coef_  must be interpreted as follows: first number is the coefficient
 # for Income, second number the coefficient for FamilySize. The estimated constant term b0 can retrieved 
 # via the .intercept_ variable.
 print(">>>Iteration ", testNumber, sep='')
 print("\tEstimated coefficients:")
 print("\t\tb1=", estimatedModel.coef_[0][0] , sep='')
 print("\t\tb2=", estimatedModel.coef_[0][1] , sep='') 
 print("\t\tb0=", estimatedModel.intercept_, sep='')
    
 predictedExpenditure = estimatedModel.predict(testData.loc[:,['temp','wind','rain']])

 # Calculate the Root Mean Squared Error (RMSE) for this testing set.
 # We have the real values of the dependent variable from the original dataset
 # which is testData.loc[:,['FoodExpenditure']] and the model's predicted values in predictedExpenditure
 # IMPORTANT: Please note the following: the function mean_squared_error() calculate the MSE NOT the RMSE. In
 # order to get the RMSE, we need to square the value returned by mean_squared_error(). See your notes on
 # how MSE and RMSE differ.
 RMSE = sqrt(mean_squared_error(testData.loc[:,['area']], predictedExpenditure))

 # Display the RMSE value
 print("\t\tModel RMSE=", RMSE, sep='')
 
 # Also, store the calculated RMSE value into an array, so that we can calculate the mean error after k-fold cross
 # validation has been finished 
 allRMSE = np.append(allRMSE, RMSE)


print("\n=======================================================")
print(" Final result: Mean RMSE of tests:", statistics.mean(allRMSE), sep='' )
print("=======================================================")

File: 7_k_fold.R
# Συνάρτηση που υπολογίζει και επιστρέφει το Μέσο Τετραγωνικό Σφάλμα (Root Mean Squared Error - RMSE)
# Επιστρέφει το Μέσο Τετραγωνικό Σφάλμα πρόβλεψης
calculateRMSE<-function(predictedValues, actualValues){
  err<- sqrt( mean((actualValues - predictedValues)^2)  )
  return( err )
}

kFoldCrossValidation<-function(data, frml, k){
  #τυχαια αναδιαταξη 
  dataset<-data[sample(nrow(data)),]
  #διαχωρισμοσ Κ κομματιων
  folds <- cut(seq(1,nrow(dataset)), breaks=k, labels=FALSE)
  #διανυσμα που αποθηκευει το rmse
  RMSE<-vector()
 
  
  # Επαναλαηπτική διαδικασία όπου κάθε ένα από τα k τμήματα θα 
  #χρησιμοποιηθεί διαδοχικά ως σύνολο ελέγχου για το μοντέλο παλινδρόμησης 
  #και όλα τα υπόλοιπα ως σύνολο εκπαίδευσης.
  # Η διαδικασία θα τερματίσει εάν όλα τα τμήματα έχουν χρησιμοποιηθεί ως σύνολο σλέγχου.
  for(i in 1:k){
    # Καθορισμός του τμήματος ελέγχου για την τρέχουσα επανάληψη 
    testIndexes <- which(folds==i,arr.ind=TRUE)
    # Καθορισμός συνόλου ελέγχου μοντέλου
    testData <- dataset[testIndexes, ]
    # Καθορισμός συνόλου εκπαίδευσης μοντέλου, που θα είναι όλα τα υπόλοιπα
    # πλην των δεδομένων που χρησιμοποιηθούν για έλεγχο
    trainData <- dataset[-testIndexes, ]
    # Εκτίμηση συντελεστών του μοντέλου παλινδρόμησης χρησιμοποιώντας το σύνολο εκπαίδευσης
    candidate.linear.model<-lm( frml, data = trainData)
    # Υπολογισμός των τιμών της εξαρτημένης μεταβλητής που προβλέπει το μοντέλο 
    # για τις τιμές του τρέχοντος συνόλου ελέγχου 
    predicted<-predict(candidate.linear.model, testData)
    # Υπολογισμός σφάλματος RMSE
    error<-calculateRMSE(predicted, testData[, "area"])
    # Αποθήκευση τιμής σφάλματος
    RMSE<-c(RMSE, error)
  }
  # Επιστροφή μέσης τιμής των σφαλμάτων που προέκυψαν απ'όλα τα τμήμα-τα ελέγχου
  return( mean(RMSE) )
}

# Ανάγνωση δεδομένων
setwd("C:/Users/user/Documents/2η μεγαλα δεδομενα/2hErgasia-2022-2023")
fires<-read.csv("forestfires.csv", sep=";", header=T)
#Έλεγχος για δεδομένα που λείπουν (missing values) και αφαίρεση ολόκληρης γραμμής
# σε περίπτωση που μια τουλάχιστον μεταβλητή έχει missing value
fires<-na.omit(fires)


# Τα τέσσερα υποψήφια μοντέλα των οποίων θα αξιολογηθεί η ικανότητα πρόβλεψης της κατανάλωσης καυσίμου
# (mpg - miles per gallon) με τη μέθοδο της διασταυρωτικής επικύρωσης 10-φορές. 
# Τα μοντέλα παλινδρόμησης αποθηκεύονται στο διάνυσμα ως συμβολοσειρές και θα μετατραπούν
# σε τύπους (formula) της R
predictionModels<-vector()
predictionModels[1]<-"area ~ temp+wind+rain"


# Μετά από κάθε διασταυρωμένη επικύρωση, ο μέσος όρος του μέσου τετρα-γωνικού 
# σφάλματος κάθε μοντέλου θα αποθηκευτεί σε διάνυσμα.
modelMeanRMSE<-vector()

# Διασταυρωμένη επικύρωση 10-φορές για κάθε ένα από τα 
# τέσσερα υποψήφια μοντέλα.
for (k in 1:length(predictionModels)){
  # Δισταυρωμένη επικύρωση 10-πτυχών για το γραμμικό μοντέλο παλιν-δρόμησης k
  modelErr<-kFoldCrossValidation(fires, as.formula(predictionModels[k]), 10)
  
  # Αποθήκευση του μέσου σφάλματος
  modelMeanRMSE<-c(modelMeanRMSE, modelErr)
  print( sprintf("Linear regression model [%s]: prediction error [%f]", predictionModels[k], modelErr ) )
}

linear.regression.model<-lm(area ~temp+wind+rain, data=fires)
print(linear.regression.model$coefficients)
summary(linear.regression.model)

####################################################################################
#ομοιως με την πανω διαδικασια αλλα τωρα για 3_2
setwd("C:/Users/user/Documents/2η μεγαλα δεδομενα/2hErgasia-2022-2023")
fires<-read.csv("forestfires.csv", sep=";", header=T)
#Έλεγχος για δεδομένα που λείπουν (missing values) και αφαίρεση ολόκληρης γραμμής
# σε περίπτωση που μια τουλάχιστον μεταβλητή έχει missing value
fires<-na.omit(fires)
fires_3_2<-subset(fires, fires$area<3.2)
linear.regression.model<-lm(area ~temp+wind+rain, data=fires_3_2)
print(linear.regression.model$coefficients)
summary(linear.regression.model)
# Επιστρέφει το Μέσο Τετραγωνικό Σφάλμα πρόβλεψης
calculateRMSE<-function(predictedValues, actualValues){
  err<- sqrt( mean((actualValues - predictedValues)^2)  )
  return( err )
}

# Συνάρτηση που υλοποιεί τον αλγόριθμο της διασταρωμένης επικύρωσης k-πτυχών.
kFoldCrossValidation<-function(data, frml, k){
  # Τυχαία αναδιάταξη 
  dataset<-data[sample(nrow(data)),]
  #Δημιουργία k
  folds <- cut(seq(1,nrow(dataset)), breaks=k, labels=FALSE)
  # Διάνυσμα όπου αποθηκεύεται το Μέσο Τετραγωικο σφαλμα
  RMSE<-vector()
  }
  
#ομοιως με τον παραπανω ελεγχο for
for(i in 1:k){
  # Καθορισμός του τμήματος ελέγχου για την τρέχουσα επανάληψη 
  testIndexes <- which(folds==i,arr.ind=TRUE)
  # Καθορισμός συνόλου ελέγχου μοντέλου
  testData <- dataset[testIndexes, ]
  # Καθορισμός συνόλου εκπαίδευσης μοντέλου, που θα είναι όλα τα υπόλοιπα
  # πλην των δεδομένων που χρησιμοποιηθούν για έλεγχο
  trainData <- dataset[-testIndexes, ]
  # Εκτίμηση συντελεστών του μοντέλου παλινδρόμησης χρησιμοποιώντας το σύνολο εκπαίδευσης
  candidate.linear.model<-lm( frml, data = trainData)
  # Υπολογισμός των τιμών της εξαρτημένης μεταβλητής που προβλέπει το μοντέλο 
  # για τις τιμές του τρέχοντος συνόλου ελέγχου 
  predicted<-predict(candidate.linear.model, testData)
  # Υπολογισμός σφάλματος RMSE
  error<-calculateRMSE(predicted, testData[, "area"])
  # Αποθήκευση τιμής σφάλματος
  RMSE<-c(RMSE, error)
}
# Επιστροφή μέσης τιμής των σφαλμάτων που προέκυψαν απ'όλα τα τμήμα-τα ελέγχου
return( mean(RMSE) )
}

predictionModels<-vector()
predictionModels[1]<-"area~temp+wind+rain"

#RMSE in a vector
modelMeanRMSE<-vector()

#10 fold cross val. for each potential model
for (k in 1:length(predictionModels)){
  modelErr<-kFoldCrossValidation(fires_3_2, as.formula(predictionModels[k]), 10)
  
  # Αποθήκευση του μέσου σφάλματος
  modelMeanRMSE<-c(modelMeanRMSE, modelErr)
  print( sprintf("Linear regression model [%s]: prediction error [%f]", predictionModels[k], modelErr ) )
}




  